### PR TITLE
[Scheduler] Proposal - initial structure for Symfony Scheduler documentation

### PR DESCRIPTION
**Pull request by alli83**
Proposal: initial structure for Symfony Scheduler documentation

This pull request serves as a proposal for the documentation plan of the Scheduler component. The objective is to make it dynamic, focusing on a  use case and exploring all the necessary setups and operational details.
The goal is to mirror the thought process one would undergo when implementing the component for a use case.

second part (6.4) https://github.com/symfony/symfony-docs/pull/19292

### CODE REVIEW

File: scheduler.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+- Compare the different cycles.
+- Introduce the Scheduler Transport and its particularity: generate directly the different messages to handle
+
+The Concept of Recurring Message in Symfony Scheduler
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A message associated with a Trigger
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- The importance of setting the frequency for your messages: Introduce the concept of a Trigger in Symfony Scheduler encompassing all the different types.
```
Comment (kbond) [inline scheduler.rst]: When covering custom triggers, an interesting example could be a trigger wrapper that excludes holidays: Cron task that runs every weekday at midnight, except "holidays" (defined by some kind of lookup table).

```php
RecurringMessage::trigger(
    new ExcludeHolidaysTrigger( // our custom trigger wrapper
        CronExpressionTrigger::fromSpec('@daily'),
    ),
    new SendDailySalesReports(),
);
```

And something like the following for the custom trigger (untested):

```php
class ExcludeHolidaysTrigger implements TriggerInterface
{
    public function __construct(private TriggerInterface $inner)
    {
    }

    public function __toString(): string
    {
        return $this->inner.' (except holidays)';
    }

    public function getNextRunDate(\DateTimeImmutable $run): ?\DateTimeImmutable
    {
        if (!$nextRun = $this->inner->getNextRunDate($run)) {
            return null;
        }

        while (!$this->isHoliday($nextRun) { // loop until we get the next run date that is not a holiday
            $nextRun = $this->inner->getNextRunDate($nextRun);
        }

        return $nextRun;
    }

    private function isHoliday(\DateTimeImmutable $timestamp): bool
    {
        // app specific logic to determine if $timestamp is on a holiday
        // returns true if holiday, false otherwise
    }
}
```

**Comment by kbond** [comment]
When documenting "hashed cron expressions", [this documentation](https://github.com/zenstruck/schedule-bundle/blob/1.x/doc/define-tasks.md#hashed-cron-expression) that explains the _spec_ can be adapted.

**Comment by alexandre-daubois** [comment]
Let's not forget about the `versionadded` directive :slightly_smiling_face: [Like this](https://github.com/TimoBakx/symfony-docs/blob/a632c9b032db8fc5632f5fb3f82a958793d7827b/webhook.rst?plain=1#L6)

File: scheduler.rst [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+------------
+
+In applications using :ref:`Symfony Flex <symfony-flex>`, run this command to
+install the scheduler component:
+
+.. code-block:: terminal
+
+    $ composer require symfony/scheduler
+
+Introduction to the case:
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
Introduction to the case
```

File: scheduler.rst [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+-------------------------
+
+Embarking on a task is one thing, but often, the need to repeat that task looms large. While one could resort to issuing commands and scheduling them with cron jobs, this approach involves external tools and additional configuration. The Symfony Scheduler component emerges as a solution, allowing us to retain control, configuration, and maintenance of task scheduling within our PHP application.
```
Comment (alexandre-daubois) [inline scheduler.rst]: As far as I am aware, we never use "us". You may replace by "allowing you to retain control". Same for "our", to be replace by "your"

File: scheduler.rst [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+At its core, the principle is straightforward: a task, considered as a message needs to be managed by a service, and this cycle must be repeated.
+Does this sound familiar? Think :doc:`Symfony Messenger docs </components/messenger>`.
```
Comment (alexandre-daubois) [inline scheduler.rst:34]: I don't know if we use that much of question-like sentences. I guess we could go for something like "This may sound familiar if you're already used to Symfony Messenger" for example

File: scheduler.rst [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+But while the system of Symfony Messenger proves very useful in various scenarios, there are instances where its capabilities fall short, particularly when dealing with repetitive tasks at regular intervals.
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
While the system of Messenger proves very useful in various scenarios, there are instances where its capabilities fall short, particularly when dealing with repetitive tasks at regular intervals.
```

File: scheduler.rst [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+Let's dive into a practical example within the context of a sales company.
+
+Imagine the company's goal is to send diverse sales reports to customers based on the specific reports each customer chooses to receive. In constructing the schedule for this scenario, the following steps are taken:
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
Let's imagine the company's goal is to send diverse sales reports to customers based on the specific reports each customer chooses to receive. In constructing the schedule for this scenario, the following steps are taken:
```

File: scheduler.rst [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+#. Iterate over reports stored in the database and create a recurring task for each report, considering its unique properties. This task, however, should not be generated during holiday periods.
+#. Furthermore, we encounter another critical task that needs scheduling: the periodic cleanup of outdated files that are no longer relevant.
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
#. The periodic cleanup of outdated files that are no longer relevant.
```

I think it's simpler yet understandable. This makes the sentence lighter :slightly_smiling_face:

File: scheduler.rst [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+On the basis of a case study in the context of a full stack Symfony application, let's dive in and explore how we can set up our system.
+
+Symfony Scheduler basics:
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
The Scheduler Basics
```

File: scheduler.rst [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+-------------------------
+
+Differences and parallels between Symfony Messenger and Symfony Scheduler.
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
Differences and Similarities Between Messenger and Scheduler
```

File: scheduler.rst [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The primary goal is to generate and process reports generation while also handling the removal of outdated reports at specified intervals.
+
+As mentioned, this component, even if it's an independent component, it draws its foundation and inspiration from the Symfony Messenger component.
```
Comment (alexandre-daubois) [inline scheduler.rst]: I'm not sure this sentence brings much added value with the one that's following

File: scheduler.rst [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+
+On one hand, it adopts well-established concepts from Symfony Messenger (such as message, handler, bus, transport, etc.). For example, the task of creating a report is considered as a message that will be directed, and processed by the corresponding handler.
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
It adopts well-established concepts from Messenger (such as message, handler, bus, transport, etc.). For example, the task of creating a report is considered as a message that will be directed, and processed by the corresponding handler.
```

You can just speak about "Messenger" and remove "Symfony". Makes it lighter :slightly_smiling_face:

File: scheduler.rst [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+    :alt: Symfony Messenger basic cycle
+
+To (Symfony Scheduler cycle):
+
+.. image:: /_images/components/scheduler/scheduler_cycle.png
+    :alt: Symfony Scheduler basic cycle
+
+In essence, it is crucial to precisely define the message to be conveyed and processed. In Symfony Scheduler, the concept of a message takes on a very particular characteristic; it should be recurrent: It's a :class:`Symfony\\Component\\Scheduler\\RecurringMessage`.
+
+Clarifying the need to define and attach Recurring Messages to a Schedule.
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
Attaching a Recurring Message To a Schedule
```

File: scheduler.rst [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to generate various messages based on their defined frequencies, configuration is necessary.
+The heart of the scheduling process and its configuration resides in a class that must extend the :class:`Symfony\\Component\\Scheduler\\ScheduleProviderInterface`.
+
+The purpose of this provider is to return a schedule through the method ``getSchedule()`` containing our different recurringMessages.
```
Comment (alexandre-daubois) [inline scheduler.rst]: You can use `:method:` here to automatically link the method to its definition on Github

File: scheduler.rst [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+
+The class attribute ``AsSchedule``, which by default references the ``default`` named schedule, allows us to register on a particular schedule::
```
Comment (alexandre-daubois) [inline scheduler.rst]: Same here for AsSchedule with `:class`:

```suggestion
The ``AsSchedule`` attribute, which by default references the ``default`` named schedule, allows us to register on a particular schedule::
```

File: scheduler.rst [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+
+    // src/Scheduler/MyScheduleProvider.php
+    namespace App\Scheduler;
+
+    class MyScheduleProvider implements ScheduleProviderInterface
+    {
+        public function getSchedule(): Schedule
+        {
+            ...
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
            // ...
```

File: scheduler.rst [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+.. tip::
+
+    This becomes important when initiating the ``messenger:consume`` command, especially when specifying one or more specific transports.
+    In Symfony Scheduler, the transport is named ``scheduler_nameofyourschedule``.
+
+The Concept of Recurring Message in Symfony Scheduler
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A message associated with a Trigger
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (alexandre-daubois) [inline scheduler.rst]: I think this title can be removed

File: scheduler.rst [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+
+The trigger is what allows configuring the recurrence frequency of your message. Several options are available to us:
+
+#. It can be a cron expression trigger:
+
+.. configuration-block::
+
+    .. code-block:: php
+
+        RecurringMessage::cron(‘*****’, new Message())
```
Comment (alexandre-daubois) [inline scheduler.rst]:
```suggestion
        RecurringMessage::cron('*****', new Message());
```

File: scheduler.rst [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+
+    * ``#yearly``, ``#annually`` - Run once a year, midnight, Jan. 1 - ``0 0 1 1 *``
+    * ``#monthly`` - Run once a month, midnight, first of month - ``0 0 1 * *``
+    * ``#weekly`` - Run once a week, midnight on Sun - ``0 0 * * 0``
+    * ``#daily``, ``#midnight`` - Run once a day, midnight - ``0 0 * * *``
+    * ``#hourly`` - Run once an hour, first minute - ``0 * * * *``
+
+.. tip::
+
+    It's possible to add and define a timezone as a 3rd argument
```
Comment (alexandre-daubois) [inline scheduler.rst]: You may actually put this in the code snippet above, something like

```
// optionally specify the timezone to run the job on
RecurringMessage::cron...

**Review by alexandre-daubois — COMMENTED** [review]
Everything can of course be discussed :smile: You should also try to break your lines at 80 characters. Congrats for this great work already :clap:

File: scheduler.rst (hunk 17 above)
Comment (kbond) [reply scheduler.rst]:
```suggestion
        RecurringMessage::cron('* * * * *', new Message())
```

(spaces between are required)

File: scheduler.rst [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+.. configuration-block::
+
+    .. code-block:: php
+
+        RecurringMessage::cron(‘*****’, new Message())
+        RecurringMessage::cron('#daily', $msg);
+
+.. tip::
+
+    A hashed expressions is a string representing the schedule for a particular command to execute.
```
Comment (kbond) [inline scheduler.rst]: This part isn't quite correct. What you're describing is the standard cron aliases

- `@daily`
- `@monthly`
- etc...

Hashed aliases are different in that they create a random but deterministic standard cron expression based on some context (in this case, the message's `__toString()` value).

At some point later in the doc, we need to discuss the importance of not having many tasks all run at the same time (ie 10 jobs at 00:00 every day). The jitter trigger wrapper is one solution, another is hashed cron expressions. I think here is where we should describe this feature.

Why don't you drop this section for now (or just show more examples of _standard_ cron expressions) and I'll create a follow-up PR explaining hashed expressions. I originally ported this code to Symfony and use it quite extensively.

File: scheduler.rst [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A message associated with a Trigger
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First and foremost, a RecurringMessage is a message that will be associated with a trigger.
+
+The trigger is what allows configuring the recurrence frequency of your message. Several options are available to us:
+
+#. It can be a cron expression trigger:
```
Comment (kbond) [inline scheduler.rst:145]: We need to note that [dragonmantank/cron-expression](https://packagist.org/packages/dragonmantank/cron-expression) is required to use the cron expression trigger.

Might also be nice to have a link to https://crontab.guru/ to help users construct/understand cron expressions.

**Comment by alli83** [comment]
I'll add a link to redirect to cache and lock

File: scheduler.rst [hunk 21]
```diff
@@ -0,0 +1,530 @@
...
+
+    $ composer require symfony/scheduler
+
+Debugging the Schedule
+~~~~~~~~~~~~~~~~~~~~~~
+
+The ``debug:scheduler`` command provides a list of schedules along with their recurring messages.
+
+You can narrow down the list to a specific schedule, or even specify a date to determine the next run date using the ``--date`` option.
+Additionally, you have the option to display terminated recurring messages using the ``--all`` option.
```
Comment (MrYamous) [inline scheduler.rst]: The `--all` option has been added in 6.4. Maybe we should keep it for another PR ?

Comment (MrYamous) [reply scheduler.rst]: Seems like `--date` has been added at the same time, 6.4 too

File: scheduler.rst [hunk 22]
```diff
@@ -0,0 +1,530 @@
...
+
+.. tip::
+
+    `dragonmantank/cron-expression`_ is required to use the cron expression trigger.
+
+    Also, `crontab_helper`_ is a good tool if you need help to construct/understand cron expressions
+
+.. tip::
+
+    It's possible to add and define a timezone as a 3rd argument
```
Comment (MrYamous) [inline scheduler.rst]: IIUC, timezone for cron expression trigger is a 6.4 feature

File: scheduler.rst [hunk 23]
```diff
@@ -0,0 +1,530 @@
...
+            yield new ReportSomethingReportSomethingElse();
+            ....
+        }
+    }
+
+Exploring alternatives for crafting your Recurring Messages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There is also another way to build a RecurringMessage, and this can be done simply by adding an attribute above a service or a command:
+:class:`Symfony\\Component\\Scheduler\\Attribute\\AsPeriodicTask` attribute and :class:`Symfony\\Component\\Scheduler\\Attribute\\AsCronTask` attribute.
```
Comment (MrYamous) [inline scheduler.rst]: Those attributes have been added in 6.4

File: scheduler.rst [hunk 24]
```diff
@@ -0,0 +1,530 @@
...
+
+In the context of our sales company, certain promotions may occur during specific periods and need to be communicated repetitively throughout a given timeframe
+or the deletion of old reports needs to be halted under certain circumstances.
+
+This is why the Scheduler incorporates a mechanism to dynamically modify the schedule and consider all changes in real-time.
+
+Exploring Strategies for adding, removing, and modifying entries within the Schedule
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The schedule provides you with the ability to :method:`Symfony\\Component\\Scheduler\Schedule::add`, :method:`Symfony\\Component\\Scheduler\Schedule::remove`, or :method:`Symfony\\Component\\Scheduler\Schedule::clear` all associated RecurringMessages,
```
Comment (MrYamous) [inline scheduler.rst]: Methods `remove` and `clear` have been added in 6.4

File: scheduler.rst [hunk 25]
```diff
@@ -0,0 +1,530 @@
...
+These events are dispatched through a listener, providing a convenient means to respond.
+
+Managing Scheduled Messages via Events
+--------------------------------------
+
+A strategic event handling
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The goal is to provide flexibility in deciding when to take action while preserving decoupling.
+Three primary event types have been introduced types
```
Comment (MrYamous) [inline scheduler.rst]: Events have been added in 6.4

**Comment by MrYamous** [comment]
Congrats for your work on this doc :)

**Comment by alli83** [comment]
@MrYamous Indeed, I've kind of overlooked the version differences 😅. I'll include them in versionadded notes
And I'll open a new PR on 6.4 for the Dynamic schedule part

File: scheduler.rst [hunk 26]
```diff
@@ -0,0 +1,328 @@
+Scheduler
+=========
+
+.. versionadded:: 6.3
+
+    The Scheduler component was introduced in Symfony 6.3
+
+The Symfony Scheduler is a powerful and flexible component designed to manage tasks scheduling within your PHP application.
```
Comment (weaverryan) [inline scheduler.rst]:
```suggestion
Scheduler is a component designed to manage task scheduling within your PHP application - like running a task each night at 3 am, every 2 weeks except for holidays or any schedule you can imagine.
```

File: scheduler.rst [hunk 27]
```diff
@@ -0,0 +1,328 @@
...
+------------
+
+In applications using :ref:`Symfony Flex <symfony-flex>`, run this command to
+install the scheduler component:
+
+.. code-block:: terminal
+
+    $ composer require symfony/scheduler
+
+Debugging the Schedule
```
Comment (weaverryan) [inline scheduler.rst]: Let's move this later. I haven't learned how to create recurring message yet.

File: scheduler.rst [hunk 28]
```diff
@@ -0,0 +1,328 @@
...
+
+Introduction to the case
+------------------------
+
+Embarking on a task is one thing, but often, the need to repeat that task looms large.
+While one could resort to issuing commands and scheduling them with cron jobs, this approach involves external tools and additional configuration.
+
+The Scheduler component emerges as a solution, allowing you to retain control, configuration, and maintenance of task scheduling within our PHP application.
+
+At its core, the principle is straightforward: a task, considered as a message needs to be managed by a service, and this cycle must be repeated.
```
Comment (weaverryan) [inline scheduler.rst]: The "considered as a message" doesn't make sense to me. Does schedule call them "messages" but the user thinks of them as tasks? If so, how about:

> At its core, scheduler allows you to create a task (called a message) that is executed by a service and repeated on some schedule.

File: scheduler.rst [hunk 29]
```diff
@@ -0,0 +1,328 @@
...
+Does this sound familiar? Think :doc:`Symfony Messenger docs </components/messenger>`.
+
+But while the system of Messenger proves very useful in various scenarios, there are instances where its capabilities
+fall short, particularly when dealing with repetitive tasks at regular intervals.
```
Comment (weaverryan) [inline scheduler.rst:37]: I think the writing - while very nice - is getting long in this section. I think (but perhaps I'm wrong) that the idea of scheduling tasks is fairly easy to understand. Pointing out that this could be done by a CRON but that scheduler allows you to keep your config all in one place & is more flexible is nice. But once we have that, we can get straight to using scheduler.

File: scheduler.rst [hunk 30]
```diff
@@ -0,0 +1,328 @@
...
+
+Let's dive into a practical example within the context of a sales company.
+
+Imagine the company's goal is to send diverse sales reports to customers based on the specific reports each customer chooses to receive.
+In constructing the schedule for this scenario, the following steps are taken:
+
+#. Iterate over reports stored in the database and create a recurring task for each report, considering its unique properties. This task, however, should not be generated during holiday periods.
+
+#. Furthermore, you encounter another critical task that needs scheduling: the periodic cleanup of outdated files that are no longer relevant.
```
Comment (weaverryan) [inline scheduler.rst:46]: I like this idea of having a real-use case that is showing non-trivial items. But I think we need some basic usage docs up higher. It's a lot of reading if I'm trying to get a general idea of what this component does.

File: scheduler.rst [hunk 31]
```diff
@@ -0,0 +1,328 @@
...
+
+From (Messenger cycle):
+
+.. image:: /_images/components/messenger/basic_cycle.png
+    :alt: Symfony Messenger basic cycle
+
+To (Scheduler cycle):
+
+.. image:: /_images/components/scheduler/scheduler_cycle.png
+    :alt: Symfony Scheduler basic cycle
```
Comment (weaverryan) [inline scheduler.rst:101]: I like these diagrams. But it may be too much for this early in the docs.

File: scheduler.rst [hunk 32]
```diff
@@ -0,0 +1,328 @@
...
+In essence, it is crucial to precisely define the message to be conveyed and processed. In Scheduler, the concept of a message takes on a very particular characteristic;
+it should be recurrent: It's a :class:`Symfony\\Component\\Scheduler\\RecurringMessage`.
+
+Clarifying the need to define and attach RecurringMessages to a Schedule
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to generate various messages based on their defined frequencies, configuration is necessary.
+The heart of the scheduling process and its configuration resides in a class that must extend the :class:`Symfony\\Component\\Scheduler\\ScheduleProviderInterface`.
+
+The purpose of this provider is to return a schedule through the method :method:`Symfony\\Component\\Scheduler\\ScheduleProviderInterface::getMethod` containing your different recurringMessages.
```
Comment (weaverryan) [inline scheduler.rst]: `getSchedule()` method, right?

File: scheduler.rst [hunk 33]
```diff
@@ -0,0 +1,328 @@
...
+
+The :class:`Symfony\\Component\\Scheduler\\Attribute\\AsSchedule` attribute, which by default references the ``default`` named schedule, allows you to register on a particular schedule::
```
Comment (weaverryan) [inline scheduler.rst:114]: Should we show the attribute below? Or is that optional?

File: scheduler.rst [hunk 34]
```diff
@@ -0,0 +1,328 @@
...
+
+    // src/Scheduler/MyScheduleProvider.php
+    namespace App\Scheduler;
+
+    class MyScheduleProvider implements ScheduleProviderInterface
```
Comment (weaverryan) [inline scheduler.rst]: If we can think of a more realistic name, that would be awesome :)

File: scheduler.rst [hunk 35]
```diff
@@ -0,0 +1,328 @@
...
+    }
+
+.. tip::
+
+    This becomes important when initiating the ``messenger:consume`` command, especially when specifying one or more specific transports.
+    In Scheduler, the transport is named ``scheduler_nameofyourschedule``.
+
+.. tip::
+
+    It is a good practice to memoize your schedule to prevent unnecessary reconstruction if the ``getSchedule`` method is checked by another service or internally within Symfony
```
Comment (weaverryan) [inline scheduler.rst:135]: I'm not sure we should mention this. Memoizing is a pretty technical term and probably creating a schedule is not a heavy thing anyway.

File: scheduler.rst [hunk 36]
```diff
@@ -0,0 +1,328 @@
...
+
+
+The Concept of RecurringMessage in Scheduler
```
Comment (weaverryan) [inline scheduler.rst]: Are there any other types of messages that you would schedule? Suggestion for the title:

> Scheduling Recurring Messages

And, I like this section. This is where we get into the meat of things

File: scheduler.rst [hunk 37]
```diff
@@ -0,0 +1,328 @@
...
+
+    // src/Scheduler/MyScheduleProvider.php
+    namespace App\Scheduler;
+
+    #[AsSchedule('uptoyou')]
+    class MyScheduleProvider implements ScheduleProviderInterface
+    {
+        public function getSchedule(): Schedule
+        {
+            return $this->schedule ??= (new Schedule())
```
Comment (weaverryan) [inline scheduler.rst:239]: Is the `$this->schedule` part of the memoization? I really need to be convinced that it is VERY important to keep this detail

File: scheduler.rst [hunk 38]
```diff
@@ -0,0 +1,328 @@
...
+                );
+        }
+    }
+
+So, this RecurringMessage will encompass both the trigger, defining the generation frequency of the message, and the message itself, the one to be processed by a specific handler.
+
+Efficient management with Symfony Scheduler
+-------------------------------------------
+
+However, if your worker becomes idle, since the messages from your schedule are generated on-the-fly by the schedulerTransport,
```
Comment (weaverryan) [inline scheduler.rst:306]: We haven't talked about workers or `messenger:consume` yet. I think we should at least show them. The story I want to see is:

A) Creating a message & handler of some sort (this can be very quick and point to Messenger, and the handler can be empty, but we should show it)
B) Creating a schedule and setting up this message in a recurring way
C) Running `messenger:consume` - and talking a bit about how that works and what to expect.

**Comment by alli83** [comment]
@weaverryan thanks for the feedback! I'll update it and ping you

File: scheduler.rst (hunk 28 above)
Comment (alli83) [reply scheduler.rst]: I understand your point of view. Indeed, what I meant to say is: 'At its core, the principle is straightforward: a task, treated as a message by the Scheduler, must be handled by a service, and this cycle should be repeated'

File: scheduler.rst (hunk 31 above)
Comment (alli83) [reply scheduler.rst:101]: Shouldn't we distinguish between the Messenger cycle and that of the Scheduler from the outset?

File: scheduler.rst (hunk 32 above)
Comment (alli83) [reply scheduler.rst]: 🙈😅

File: scheduler.rst (hunk 33 above)
Comment (alli83) [reply scheduler.rst:114]: oh yes indeed, it's just an oversight. If you don't specify the name (just AsSchedule attribute), it will default to 'default'. The way it's written in the tip section can be misleading: I update it

File: scheduler.rst [hunk 39]
```diff
@@ -0,0 +1,357 @@
...
+
+.. tip::
+
+    `dragonmantank/cron-expression`_ is required to use the cron expression trigger.
+
+    Also, `crontab_helper`_ is a good tool if you need help to construct/understand cron expressions
+
+.. versionadded:: 6.4
+
+    Since version 6.4, it is now possible to add and define a timezone as a 3rd argument
```
Comment (MrYamous) [inline scheduler.rst:161]:
```suggestion
```

To keep for 6.4 PR

File: scheduler.rst [hunk 40]
```diff
@@ -0,0 +1,357 @@
...
+Debugging the Schedule
+~~~~~~~~~~~~~~~~~~~~~~
+
+The ``debug:scheduler`` command provides a list of schedules along with their recurring messages.
+You can narrow down the list to a specific schedule.
+
+.. versionadded:: 6.4
+
+    Since version 6.4, you can even specify a date to determine the next run date using the ``--date`` option.
+    Additionally, you have the option to display terminated recurring messages using the ``--all`` option.
```
Comment (MrYamous) [inline scheduler.rst:284]:
```suggestion
```

File: scheduler.rst (hunk 39 above)
Comment (alli83) [reply scheduler.rst:161]: https://github.com/symfony/symfony-docs/pull/19244#issuecomment-1851974661 
It'll be updated in https://github.com/symfony/symfony-docs/pull/19292 + version added in 6.3

File: scheduler.rst [hunk 41]
```diff
@@ -0,0 +1,378 @@
...
+    $ composer require symfony/scheduler
+
+
+Introduction to the case
+------------------------
+
+Embarking on a task is one thing, but often, the need to repeat that task looms large.
+While one could resort to issuing commands and scheduling them with cron jobs, this approach involves external tools and additional configuration.
+
+The Scheduler component emerges as a solution, allowing you to retain control, configuration, and maintenance of task scheduling within our PHP application.
```
Comment (fabpot) [inline scheduler.rst:31]:
```suggestion
The Scheduler component emerges as a solution, allowing you to retain control, configuration, and maintenance of task scheduling within PHP applications.
```

File: scheduler.rst [hunk 42]
```diff
@@ -0,0 +1,378 @@
...
+
+At its core, scheduler allows you to create a task (called a message) that is executed by a service and repeated on some schedule.
```
Comment (fabpot) [inline scheduler.rst:33]:
```suggestion
At its core, the Scheduler component allows you to create a task (called a message) that is executed by a service and repeated on some schedule.
```

File: scheduler.rst [hunk 43]
```diff
@@ -0,0 +1,378 @@
...
+.. code-block:: terminal
+
+    $ composer require symfony/scheduler
+
+
+Introduction to the case
+------------------------
+
+Embarking on a task is one thing, but often, the need to repeat that task looms large.
+While one could resort to issuing commands and scheduling them with cron jobs, this approach involves external tools and additional configuration.
```
Comment (fabpot) [inline scheduler.rst:29]: I think the main benefit is that everything is managed in your app and that gives a lot of flexibility that is just not possible with cron jobs (dynamic schedules being the main one).

Comment (alli83) [reply scheduler.rst:29]: yes, indeed, I should emphasize this point more. I'll update

**Comment by javiereguiluz** [comment]
@alli83 we've finally decided to merge your PR "as is". In a different PR we're going to make some of the changes proposed by the reviewers.

Thanks a lot for your contribution to kickstart these docs and thanks to all reviewers too.

**Comment by alexandre-daubois** [comment]
Congrats @alli83!

**Comment by alli83** [comment]
> @alli83 we've finally decided to merge your PR "as is". In a different PR we're going to make some of the changes proposed by the reviewers.
> 
> Thanks a lot for your contribution to kickstart these docs and thanks to all reviewers too.

Ok thx for all your feedback. second part (6.4) is here https://github.com/symfony/symfony-docs/pull/19292
